Download content disposition

mgrauer opened the pull request:
This allows setting the content-disposition on a download of a single file.

This works for all files except for anonymously accessed files stored in S3 assetstores, due to a limitation from Amazon.  S3 allows you to specify the content-disposition parameter of a download URL, except in this case.

From the S3 [GET Object documentation](http://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectGET.html)

>You must sign the request, either using an Authorization header or a pre-signed URL, when using these parameters. They cannot be used with an unsigned (anonymous) request.

I've whitelisted two values for Content-Disposition at the REST layer, "inline" and "attachment", to address @brianhelba 's concerns.  


Still need to add tests on this PR.

Fixes #1118.

mgrauer commented:
- [ ] add tests
- [ ] add filename to content-disposition response

mgrauer commented:
Ready for review.

If we decide to keep the whitelist of content-disposition for item and file download endpoints, we should unify it somewhere.  in girder.constants?

    allowed = ['inline', 'attachment']

zachmullen commented:
Whoops, I broke this... can you merge master

mgrauer commented:
merged

zachmullen commented on girder/api/v1/file.py:
```diff
@@ -244,6 +244,8 @@ def readChunk(self, params):
                'final byte you wish to receive.', dataType='integer',
                required=False)
+        .param('contentDisposition', 'Specify the Content-Disposition response '
+               'header disposition-type value', required=False)
```
We should use the `enum` kwarg to this function to enumerate the allowed strings.

zachmullen commented on girder/api/v1/file.py:
```diff
@@ -266,7 +268,15 @@ def download(self, file, params):
+                                contentDisp)
+
+        return \
+            self.model('file').download(file, offset, endByte=endByte,
```
To avoid backslash here you can break after the opening paren `download(` and just indent the next line.

mgrauer commented:
@zachmullen 

I lost track of this, but have now addressed your comments.

brianhelba commented:
@mgrauer So I actually needed this branch last night, and I had to rebase it to my local deployment. I'd like to see it merged.

I have a few code comments though, I'll put them in.

brianhelba commented on girder/api/v1/file.py:
```diff
@@ -244,6 +244,9 @@ def readChunk(self, params):
                'final byte you wish to receive.', dataType='integer',
                required=False)
+        .param('contentDisposition', 'Specify the Content-Disposition response '
+               'header disposition-type value', required=False,
```
You can specify a ``default`` value here. Even though the API layer just sets it to ``None``, internally it will always default to ``attachment``.

brianhelba commented on girder/api/v1/file.py:
```diff
@@ -266,7 +269,14 @@ def download(self, file, params):
 
-        return self.model('file').download(file, offset, endByte=endByte)
+        contentDisp = params.get('contentDisposition', None)
+        allowed = ['inline', 'attachment']
```
``allowed`` should be a set (just define the literal inside curly braces). You can just inline the variable too.

brianhelba commented on girder/utility/s3_assetstore_adapter.py:
```diff
@@ -392,7 +392,13 @@ def downloadFile(self, file, offset=0, headers=True, endByte=None,
+                    # Tell S3 to set Content-Disposition response header,
+                    # though this only works for non-anonymous connections.
+                    queryParams['response-content-disposition'] = \
+                        'inline; filename="%s"' % file['name']
```
I assume S3 just defaults to ``attachment``? A code comment here noting that fact would be good.

mgrauer commented:
@brianhelba 

Girder is setting the Content-Disposition for S3 files on the initial [upload](https://github.com/girder/girder/blob/53a39f77892b13b4a7aae0640694d868e99c465d/girder/utility/s3_assetstore_adapter.py#L177-L178) to be "attachment; filename={$filename}" and then can be changed on any [updates](https://github.com/girder/girder/blob/53a39f77892b13b4a7aae0640694d868e99c465d/girder/utility/s3_assetstore_adapter.py#L501-L508).

I've addressed your comments.

brianhelba commented:
LGTM.

brianhelba commented:
LGTM.
